src/diem/miniwallet/app/app.py

```diff
@@ -0,0 +1,263 @@
+            raise ValueError("%r value must be greater than or equal to zero" % name)
+
+    def _validate_account_balance(self, txn: Transaction) -> None:
+        if txn.payee:
+            balance = self._balance(str(txn.account_id), txn.currency)
+            if balance < txn.amount:
+                raise ValueError("account balance not enough: %s < %s" % (balance, txn.amount))
+
+    def _balance(self, account_id: str, currency: str) -> int:
+        txns = self.store.find_all(Transaction, account_id=account_id, currency=currency)
+        return sum([t.balance_amount() for t in txns if t.status != "canceled"])
+
+    def _gen_subaddress(self) -> bytes:
+        self.subaddress_id += 1
+        return self.subaddress_id.to_bytes(8, byteorder="big")
+
+    def _to_offchain_command(self, cmd: Command) -> offchain.Command:
+        req = cmd.request_object()
+        if isinstance(req.command, PaymentCommandObject):
+            return self.offchain.create_payment_command(req.cid, req.command.payment)
+        raise ValueError("unsupported command type: %s" % req.command_type)
+
+    def _txn_metadata(self, txn: Transaction) -> Tuple[bytes, bytes]:
+        if txn.reference_id:
+            cmd = self.store.find(Command, last=True, reference_id=txn.reference_id)
+            return self.diem.travel_metadata(cast(PaymentCommand, self._to_offchain_command(cmd)))
+        if txn.subaddress_hex:
+            return self.diem.general_metadata(txn.subaddress(), str(txn.payee))
+        raise ValueError("transaction is not ready for submit diem transaction: %s" % txn)
+
+
+class OffChainAPI(Base):
+    def offchain_api(self, request_sender_address: str, request_bytes: bytes) -> CommandResponseObject:
+        cmd = self.offchain.process_inbound_request(request_sender_address, request_bytes)
+        with self.store.transaction:
+            getattr(self, "_handle_offchain_%s" % utils.to_snake(cmd))(cmd)
+        return offchain.reply_request(cid=cmd.id())
+
+    def jws_serialize(self, resp: CommandResponseObject) -> bytes:
+        return offchain.jws.serialize(resp, self.diem.account.compliance_key.sign)
+
+    def _handle_offchain_payment_command(self, new_cmd: PaymentCommand) -> None:
+        try:
+            cmd = self.store.find(Command, last=True, reference_id=new_cmd.reference_id())
+            payment_command = self._to_offchain_command(cmd)
+            account_id = cmd.account_id
+        except NotFoundError:
+            payment_command = None
+            subaddress = utils.hex(new_cmd.my_subaddress(self.diem.account.hrp))
+            account_id = self.store.find(PaymentURI, subaddress_hex=subaddress).account_id
+
+        if new_cmd != payment_command:
+            new_cmd.validate(payment_command)
+            self._create_account_command(account_id, new_cmd)
+
+    def _process_offchain_commands(self) -> None:
+        cmds = {cmd.reference_id: cmd for cmd in self.store.find_all(Command)}
+        for cmd in cmds.values():
```

Comment:
`for cmd in self.store.find_all(Command):` seems to be sufficient (building up the `cmds` dict above seems redundant)

Reply:
It is because for same reference_id, there are multiple Command, if you noticed how to find Command by reference_id, it has last=True, and we only create command and never update existing command by reference id.
One reason to store all commands, is for test to fetch all commands for a given reference_id, so that we know what is off-chain communication steps.